### PR TITLE
Allow adding tags to a conversation upon creation

### PR DESCRIPTION
**Pull request by swarelle**
#52 
Added another input box on conversations.jsp's "New Conversation" section to allow the user to add tags to the conversation upon creation.

### CODE REVIEW

File: src/main/java/codeu/controller/ConversationServlet.java [hunk 1]
```diff
@@ -116,6 +119,23 @@ public void doPost(HttpServletRequest request, HttpServletResponse response)
...
+    String tags = request.getParameter("tags");
+    if (tags != null) {
+      if (!tags.matches("([\\w*](, ))*[\\w*]")) {
+        request.setAttribute("error", "Please enter tags as comma-separated words with one space between them. Tags can only contain letters and numbers.");
+        request.getRequestDispatcher("/WEB-INF/view/conversations.jsp").forward(request, response);
+        return;
+      }
+
+      // Add all tags to the user's tagged conversations Map
+      List<String> splitTags = Arrays.asList(tags.split(", "));
```
Comment (joycezhao2) [inline src/main/java/codeu/controller/ConversationServlet.java]: do we want to make tags case sensitive? e.g. tag "Work" is the same as the tag "work"

File: src/main/java/codeu/controller/ConversationServlet.java [hunk 2]
```diff
@@ -15,11 +15,14 @@
 package codeu.controller;
 
 import codeu.model.data.Conversation;
+import codeu.model.data.Tag;
 import codeu.model.data.User;
 import codeu.model.store.basic.ConversationStore;
 import codeu.model.store.basic.UserStore;
+import com.google.appengine.repackaged.com.google.api.client.util.Lists;
```
Comment (joycezhao2) [inline src/main/java/codeu/controller/ConversationServlet.java]: What is this used for?

File: src/main/java/codeu/model/data/User.java [hunk 3]
```diff
@@ -56,8 +65,33 @@ public User(UUID id, String name, Instant creation, String hashedPassword) {
     this.creation = creation;
     this.hashedPassword = hashedPassword;
     this.about = "";
+    this.taggedConversations = new HashMap<>();
+    this.conversationTags = new HashMap<>();
   }
 
+  /**
+   * Adds a Tag to a Conversation.
```
Comment (joycezhao2) [inline src/main/java/codeu/model/data/User.java]: Are theses conversations that a user has created themselves?

File: src/main/java/codeu/controller/ConversationServlet.java [hunk 4]
```diff
@@ -116,6 +119,23 @@ public void doPost(HttpServletRequest request, HttpServletResponse response)
     Conversation conversation =
         new Conversation(UUID.randomUUID(), user.getId(), conversationTitle, Instant.now());
 
+    // Get all tags
+    String tags = request.getParameter("tags");
```
Comment (joycezhao2) [inline src/main/java/codeu/controller/ConversationServlet.java]: Where are we actually going to set the tag parameter for the request?

**Comment by joycezhao2** [comment]
also can you edit the title/description with more details about the PR?

File: src/main/java/codeu/controller/ConversationServlet.java (hunk 1 above)
Comment (swarelle) [reply src/main/java/codeu/controller/ConversationServlet.java]: i think case insensitive probably makes more sense -- will update equals() to reflect that.

File: src/main/java/codeu/controller/ConversationServlet.java (hunk 2 above)
Comment (swarelle) [reply src/main/java/codeu/controller/ConversationServlet.java]: Good question! It was an import Eclipse did automatically...

File: src/main/java/codeu/model/data/User.java (hunk 3 above)
Comment (swarelle) [reply src/main/java/codeu/model/data/User.java]: Nope, these are the Conversations in ConversationStore

Comment (joycezhao2) [reply src/main/java/codeu/model/data/User.java]: I'm a little confused, when will we call this method?

File: src/main/java/codeu/controller/ConversationServlet.java [hunk 5]
```diff
@@ -116,6 +118,23 @@ public void doPost(HttpServletRequest request, HttpServletResponse response)
...
+        request.setAttribute("error", "Please enter tags as comma-separated words with one space between them. Tags can only contain letters and numbers.");
+        request.getRequestDispatcher("/WEB-INF/view/conversations.jsp").forward(request, response);
+        return;
+      }
+
+      // Add all tags to the user's tagged conversations Map
+      List<String> splitTags = Arrays.asList(tags.split(", "));
+      for (String tag : splitTags) {
+        Tag newTag = new Tag(UUID.randomUUID(), tag, Instant.now());
+        user.addTaggedConversation(newTag, conversation);
```
Comment (swarelle) [inline src/main/java/codeu/controller/ConversationServlet.java]: Changes to make:
1. Check if Tag exists in TagStore
2. Add Tag to Conversation and add Conversation to Tag.

File: src/main/java/codeu/model/data/User.java [hunk 6]
```diff
@@ -15,6 +15,11 @@
 package codeu.model.data;
 
 import java.time.Instant;
+import java.util.HashMap;
```
Comment (swarelle) [inline src/main/java/codeu/model/data/User.java]: Don't need User.java changes anymore

**Comment by swarelle** [comment]
Closing to restructure.

**Pull request closed without merging**